### PR TITLE
Update the clouddriver lambda APIs to support additional attributes and Update the load balancer to support lambda target.

### PR DESCRIPTION
**Pull request by madhusarma**
**Issue Summary**:

Update the clouddriver lambda APIs to support additional attributes and Update the load balancer to support lambda target.

**Cloud Provider(s):**

AWS

**Environment**:

**Feature Area :**   Clouddriver

**Description**:
1) Update  "UpdateLambdaConfigurationAtomicOperation" to support the VPC, tags,environment variables, etc.
2) Update CreateLambdaAtomicOperation to add VPC, tags,environment variables, while creating the Lambda.
3) Add Lambda relation to Spinnaker application.
4) Add Lambda support to loadbalancer target groups.

### CODE REVIEW

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/data/Keys.groovy [hunk 1]
```diff
@@ -175,6 +176,10 @@ class Keys implements KeyParser {
   }
 
   static String getTargetGroupKey(String targetGroupName, String account, String region, String targetGroupType, String vpcId) {
+    //Lambda targetGroup don't have the vpcId
+    if (TargetTypeEnum.Lambda.toString().equalsIgnoreCase(targetGroupType))
+     "${ID}:${Namespace.TARGET_GROUPS}:${account}:${region}:${targetGroupName}:${targetGroupType}"
+    else
```
Comment (jeyrschabu) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/data/Keys.groovy]: style nit: prefer a fully enclosed `if/else`

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/handlers/LoadBalancerV2UpsertHandler.groovy [hunk 2]
```diff
@@ -42,24 +42,31 @@ class LoadBalancerV2UpsertHandler {
...
-        targetGroupAttributes.add(new TargetGroupAttribute(key: "deregistration_delay.timeout_seconds", value: deregistrationDelay.toString()))
-      }
-      if (loadBalancer.type == 'application') {
-        if (attributes.stickinessEnabled != null) {
-          targetGroupAttributes.add(new TargetGroupAttribute(key: "stickiness.enabled", value: attributes.stickinessEnabled.toString()))
+      if(TargetTypeEnum.Lambda.toString().equalsIgnoreCase(targetGroup.getTargetType()))
+      {
+        if (attributes.multiValueHeadersEnabled != null) {
+          targetGroupAttributes.add(new TargetGroupAttribute(key: "lambda.multi_value_headers.enabled", value: attributes.multiValueHeadersEnabled))
         }
```
Comment (jeyrschabu) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/handlers/LoadBalancerV2UpsertHandler.groovy]: These conditionals can be collapsed to a single line: Checking if the target type is lambda and that the `multiValueHeadersEnabled` is set.

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/handlers/LoadBalancerV2UpsertHandler.groovy [hunk 3]
```diff
@@ -42,24 +42,31 @@ class LoadBalancerV2UpsertHandler {
...
+          if (attributes.stickinessDuration != null) {
+            targetGroupAttributes.add(new TargetGroupAttribute(key: "stickiness.lb_cookie.duration_seconds", value: attributes.stickinessDuration.toString()))
+          }
         }
-      }
-      if(loadBalancer.type == 'network' ){
-        if(attributes.proxyProtocolV2 != null){
-          targetGroupAttributes.add(new TargetGroupAttribute(key: "proxy_protocol_v2.enabled", value: attributes.proxyProtocolV2))
+        if(loadBalancer.type == 'network' ){
+          if(attributes.proxyProtocolV2 != null){
```
Comment (jeyrschabu) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/handlers/LoadBalancerV2UpsertHandler.groovy]: nit: some needed spaces around `if`

File: clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/provider/agent/LambdaCachingAgent.java [hunk 4]
```diff
@@ -212,4 +248,152 @@ public CacheResult loadData(ProviderCache providerCache) {
...
+                          targetGroupNames.add(name.get());
+                      }
+                    }
+                  }
+                }
+              }
+            }
+          }
+        }
+      }
```
Comment (jeyrschabu) [inline clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/provider/agent/LambdaCachingAgent.java]: This nesting takes a hit on readability

File: clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/provider/agent/LambdaCachingAgent.java [hunk 5]
```diff
@@ -212,4 +248,152 @@ public CacheResult loadData(ProviderCache providerCache) {
...
+  }
+
+  private List<String> getTargetGroupNames(AWSLambda lambda, String functionName) {
+    List<String> targetGroupNames = new ArrayList<>();
+    try {
+      GetPolicyResult result =
+          lambda.getPolicy(new GetPolicyRequest().withFunctionName(functionName));
+      String json = result.getPolicy();
+      Policy policy = Policy.fromJson(json);
+      for (Statement statement : policy.getStatements()) {
```
Comment (cfieber) [inline clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/provider/agent/LambdaCachingAgent.java]: as I commented on the previous incarnation of this PR - I'd like to see this flattened out via the streams api

File: clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/provider/agent/LambdaCachingAgent.java [hunk 6]
```diff
@@ -212,4 +248,152 @@ public CacheResult loadData(ProviderCache providerCache) {
...
+                  }
+                }
+              }
+            }
+          }
+        }
+      }
+    } catch (ResourceNotFoundException ex) {
+      // ignore the exception.
+      log.info("No policies are exist for {}", functionName);
```
Comment (cfieber) [inline clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/provider/agent/LambdaCachingAgent.java]:
```suggestion
      log.info("No policies exist for {}", functionName);
```

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/data/Keys.groovy (hunk 1 above)
Comment (madhusarma) [reply clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/data/Keys.groovy]: Implemented.

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/handlers/LoadBalancerV2UpsertHandler.groovy (hunk 2 above)
Comment (madhusarma) [reply clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/handlers/LoadBalancerV2UpsertHandler.groovy]: The reason for this is, first "if" has "else", if we combine two "if" statements, then we have to replace "else" with "if". Please let me know your thoughts.
```
if(condition1){
       if(condition2){
        }
}else{
......
}
```

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/handlers/LoadBalancerV2UpsertHandler.groovy (hunk 3 above)
Comment (madhusarma) [reply clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/handlers/LoadBalancerV2UpsertHandler.groovy]: Implemented.

File: clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/provider/agent/LambdaCachingAgent.java (hunk 4 above)
Comment (madhusarma) [reply clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/provider/agent/LambdaCachingAgent.java]: Replaced this with Stream API.

File: clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/model/FunctionProvider.java [hunk 7]
```diff
@@ -16,10 +16,23 @@
...
+  /**
+   * Returns all functions related to an application based on one of the following criteria: - the
+   * load balancer name follows the Frigga naming conventions for load balancers (i.e., the load
+   * balancer name starts with the application name, followed by a hyphen)
+   *
+   * @param applicationName the name of the application
+   * @return a collection of functions.
+   */
+  @Empty
+  Set<? extends Function> getApplicationFunctions(String applicationName);
```
[the remark this answers is not in the document]
Comment (madhusarma) [reply clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/model/FunctionProvider.java]: Added the default implementation.

File: clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/provider/agent/LambdaCachingAgent.java (hunk 5 above)
Comment (madhusarma) [reply clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/provider/agent/LambdaCachingAgent.java]: Replaced this with Streams API.

File: clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/provider/agent/LambdaCachingAgent.java (hunk 6 above)
Comment (madhusarma) [reply clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/provider/agent/LambdaCachingAgent.java]: Corrected.

**Comment by madhusarma** [comment]
Implemented the required changes as per the commits, Please review and approve.

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/handlers/LoadBalancerV2UpsertHandler.groovy [hunk 8]
```diff
@@ -42,24 +42,31 @@ class LoadBalancerV2UpsertHandler {
...
+          }
+          if (attributes.stickinessDuration != null) {
+            targetGroupAttributes.add(new TargetGroupAttribute(key: "stickiness.lb_cookie.duration_seconds", value: attributes.stickinessDuration.toString()))
+          }
         }
-      }
-      if(loadBalancer.type == 'network' ){
-        if(attributes.proxyProtocolV2 != null){
-          targetGroupAttributes.add(new TargetGroupAttribute(key: "proxy_protocol_v2.enabled", value: attributes.proxyProtocolV2))
+        if(loadBalancer.type == 'network' ) {
```
[the remark this answers is not in the document]
Comment (madhusarma) [reply clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/handlers/LoadBalancerV2UpsertHandler.groovy]: Implemented.

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/handlers/LoadBalancerV2UpsertHandler.groovy [hunk 9]
```diff
@@ -42,24 +42,31 @@ class LoadBalancerV2UpsertHandler {
...
     def targetGroupAttributes = []
     if (attributes) {
-      Integer deregistrationDelay = [attributes.deregistrationDelay, deployDefaults?.loadBalancing?.deregistrationDelayDefault].findResult(Closure.IDENTITY)
-      if (deregistrationDelay != null) {
-        targetGroupAttributes.add(new TargetGroupAttribute(key: "deregistration_delay.timeout_seconds", value: deregistrationDelay.toString()))
-      }
-      if (loadBalancer.type == 'application') {
-        if (attributes.stickinessEnabled != null) {
-          targetGroupAttributes.add(new TargetGroupAttribute(key: "stickiness.enabled", value: attributes.stickinessEnabled.toString()))
+      if(TargetTypeEnum.Lambda.toString().equalsIgnoreCase(targetGroup.getTargetType()))
```
[the remark this answers is not in the document]
Comment (madhusarma) [reply clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/handlers/LoadBalancerV2UpsertHandler.groovy]: Implemented.

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/handlers/LoadBalancerV2UpsertHandler.groovy (hunk 8 above)
[the remark this answers is not in the document]
Comment (madhusarma) [reply clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/handlers/LoadBalancerV2UpsertHandler.groovy]: Implemented the PR comments, Please review and approve.

**Comment by madhusarma** [comment]
Implemented the PR comments, Please review and approve.

**Comment by madhusarma** [comment]
@cfieber Implemented the required changes,  please review.

**Pull request merged**